### PR TITLE
Refactor user and teams loading into hook

### PR DESCRIPTION
**Pull request by selfcontained**
## Description
Starting some cleanup on the top level `<App/>` component to make it a bit easier to follow/maintain. The main task in this PR was to pull the `useEffect()` statement that loads the user and teams into it's own hook (I created a new `hooks` folder as well) to try and encapsulate some of the core data that needs to load for the app to render. I've also left a few comments in there for followup tasks and clarifications.

## How to test
**Preview Env:** https://bmh-dashbocf42e343d3.preview.gitpod-dev.com/

Since this is the top level component, loading the app on the preview env should exercise the changes.

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@selfcontained: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/user-context.tsx [hunk 1]
```diff
@@ -22,9 +22,11 @@ const UserContext = createContext<{
 const UserContextProvider: React.FC = ({ children }) => {
     const [user, setUser] = useState<User>();
     const [billingMode, setBillingMode] = useState<BillingMode>();
-    function refreshUserBillingMode() {
-        getGitpodService().server.getBillingModeForUser().then(setBillingMode);
-    }
+
+    const refreshUserBillingMode = useCallback(() => {
```
Comment (selfcontained) [inline components/dashboard/src/user-context.tsx:26]: Wrapping in a `useCallback()` so we don't create a new reference on re-renders

**Comment by easyCZ** [comment]
Looks great. I'll give it a more thorough pass on Monday but very nice to see this!

File: components/dashboard/src/hooks/use-user-and-teams-loader.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                trackLocation(!!loggedInUser);
+            }
+            setLoading(false);
+            (window as any)._gp.path = window.location.pathname; //store current path to have access to previous when path changes
+        })();
+        // Ensure this only ever runs once
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, []);
+
+    // TODO: Can this check happen when we load the teams rather than a separate effect?
```
Comment (svenefftinge) [inline components/dashboard/src/hooks/use-user-and-teams-loader.ts:73]: 👍

File: components/dashboard/src/hooks/use-user-and-teams-loader.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    useEffect(() => {
+        if (!teams) {
+            return;
+        }
+        // Refresh billing mode (side effect on other components per UserContext!)
+        refreshUserBillingMode();
+    }, [teams, refreshUserBillingMode]);
+
+    // TODO: Can this check happen when we load the user rather than a separate effect?
```
Comment (svenefftinge) [inline components/dashboard/src/hooks/use-user-and-teams-loader.ts:82]: yes

**Review by svenefftinge — APPROVED** [review]
LGTM

/hold

**Review by easyCZ — APPROVED** [review]

**Pull request merged**